Tools - Profiling gdoc to md

Ro0k1e opened the pull request:
Issue [#428](https://github.com/sorrentum/sorrentum/issues/428)

samarth9008 commented:
Also add @DanilYachmenev as a reviewer. Assign this PR to yourself.

samarth9008 commented on docs/Tools_Profiling.md:
```diff
@@ -0,0 +1,367 @@
+# Profiling end-to-end a command line
```
You add name of the file over here which is `Tools - Profiling`. This will be first in TOC.

samarth9008 commented on docs/Tools_Profiling.md:
```diff
@@ -0,0 +1,367 @@
+
+<!-- toc -->
+
+- [Profiling Python code from command line](#profiling-python-code-from-command-line)
+  * [cProfile](#cprofile)
+    + [Install in a Docker container](#install-in-a-docker-container)
+    + [How to use with workflow](#how-to-use-with-workflow)
+    + [How to use manually](#how-to-use-manually)
+    + [process_prof.py](#process_profpy)
+  * [line_profiler](#line_profiler)
+    + [How to use](#how-to-use)
+  * [pytest-profiling](#pytest-profiling)
+    + [How to use](#how-to-use-1)
+- [Profiling in a Jupyter notebook](#profiling-in-a-jupyter-notebook)
+  * [Time profilers](#time-profilers)
+  * [By function](#by-function)
+- [Memory profilers](#memory-profilers)
```
The index should look like this. Modify the heading in the doc and run the linter again to generate the TOC on its own.
![image](https://github.com/sorrentum/sorrentum/assets/58821247/5720ccc7-8d04-4573-91db-da61fe03accf)

samarth9008 commented on docs/Tools_Profiling.md:
````diff
@@ -0,0 +1,367 @@
+
+Elapsed (wall clock) time (h:mm:ss or m:ss): 3:31.38
+
+Average shared text size (kbytes): 0
+
+Average unshared data size (kbytes): 0
+
+Average stack size (kbytes): 0
+
+Average total size (kbytes): 0
+
+Maximum resident set size (kbytes): 13083892
+
+Average resident set size (kbytes): 0
+
+Major (requiring I/O) page faults: 0
+
+Minor (reclaiming a frame) page faults: 9911066
+
+Voluntary context switches: 235772
+
+Involuntary context switches: 724
+
+Swaps: 0
+
+File system inputs: 424
+
+File system outputs: 274320
+
+Socket messages sent: 0
+
+Socket messages received: 0
+
+Signals delivered: 0
+
+Page size (bytes): 4096
+
+Exit status: 0
+
+```
````
No need for this extra spaces in between inside the code format. Improve it everywhere.

samarth9008 commented on docs/Tools_Profiling.md:
```diff
@@ -0,0 +1,367 @@
+
+There is a script that runs the flow
+
+amp/dev_scripts/run_profiling.sh
```
Lets add this in code format.

samarth9008 commented on docs/Tools_Profiling.md:
```diff
@@ -0,0 +1,367 @@
+# Profiling in a Jupyter notebook
+
+You can find all of the examples below in action in the
+time_memory_profiling_example.ipynb notebook.
```
time_memory_profiling_example.ipynb -> `time_memory_profiling_example.ipynb`

samarth9008 commented on docs/Tools_Profiling.md:
```diff
@@ -0,0 +1,367 @@
+
+This will output something like this:
+
+<img src="Tools_Profiling_figs/image1.png" style="width:6.5in;height:12.5in" />
```
Both images are not displayed

DanilYachmenev commented on docs/Tools_Profiling.md:
```diff
@@ -0,0 +1,274 @@
+We prefer cProfile for profiling and gprof2dot for visualization.
+
+The documentation does not state this, but
+`%prun magic uses cProfile under the hood, so we can use it in the notebook instead: `
```
put only `%prun` in backticks so the text following it is not formatted as code

gpsaggese commented:
Not sure I understand what happened here. Why was the PR closed?

DanilYachmenev commented:
@Ro0k1e just a reminder for further work - you should never merge changes to the master by yourself ok?
Only integrators are responsible for this
Just in case

Ro0k1e commented:
> @Ro0k1e just a reminder for further work - you should never merge changes to the master by yourself ok? Only integrators are responsible for this Just in case

Sure! Thank you so much for the reminder and sorry for the confusion. What I meant was a pull request.

gpsaggese commented:
> > Not sure I understand what happened here. Why was the PR closed?
> 
> Hi GP!
> 
> To deal with issue #428, I created a branch [SorrIssue428_Tools_Profiling_gdoc_to_md](https://github.com/sorrentum/sorrentum/tree/SorrIssue428_Tools_Profiling_gdoc_to_md) to be merged to the master and this PR. Yesterday I found that there were some changes to the master, and the branch was not up-to-date. I was afraid that merging the branch to the master might introduce some outdated files to the master, so I closed this PR, deleted the old branch, and created a new one with the same name based on the updated master. The new PR can be found here: #452.

I understand
1) Git allows you to do all these things seamlessly, without creating new branches. E.g., you could have just merge master and accept your branch in case of conflicts
- Pls put some time in studying https://github.com/gpsaggese/umd_data605/blob/main/lectures/02.1%20-%20Git%2C%20Data%20Pipelines.pdf and the Git book. It will pay 10x
2) You can add an explanation (like this) in the PR before closing. Otherwise people wonder what the heck happened? (principle of least surprise)

@DanilYachmenev I've updated the merging rules so we should not have rogue merges. Let's monitor and improve the rules for a pacific coexistence.

Ro0k1e commented:
> > > Not sure I understand what happened here. Why was the PR closed?
> > 
> > 
> > Hi GP!
> > To deal with issue #428, I created a branch [SorrIssue428_Tools_Profiling_gdoc_to_md](https://github.com/sorrentum/sorrentum/tree/SorrIssue428_Tools_Profiling_gdoc_to_md) to be merged to the master and this PR. Yesterday I found that there were some changes to the master, and the branch was not up-to-date. I was afraid that merging the branch to the master might introduce some outdated files to the master, so I closed this PR, deleted the old branch, and created a new one with the same name based on the updated master. The new PR can be found here: #452.
> 
> I understand
> 
> 1. Git allows you to do all these things seamlessly, without creating new branches. E.g., you could have just merge master and accept your branch in case of conflicts
> 
> * Pls put some time in studying https://github.com/gpsaggese/umd_data605/blob/main/lectures/02.1%20-%20Git%2C%20Data%20Pipelines.pdf and the Git book. It will pay 10x
> 
> 2. You can add an explanation (like this) in the PR before closing. Otherwise people wonder what the heck happened? (principle of least surprise)
> 
> @DanilYachmenev I've updated the merging rules so we should not have rogue merges. Let's monitor and improve the rules for a pacific coexistence.

Understood! Thank you so much for the help!
